spikeinterface/sortingcomponents/peak_detection.py

```diff
@@ -347,3 +405,152 @@ def _numba_detect_peak_neg(traces, traces_center, peak_mask, exclude_sweep_size,
+            )
+        elif peak_sign == "pos":
+            neg_traces = torch.as_tensor(
+                traces, device=device, dtype=torch.float
+            )
+        elif peak_sign == "both":
+            neg_traces = torch.as_tensor(
+                -np.abs(traces), device=device, dtype=torch.float
+            )
+        thresholds_torch = torch.as_tensor(thresholds, device=device, dtype=torch.float)
+
+        max_amps, inds = F.max_pool2d_with_indices(
+            neg_traces[None, None],
+            kernel_size=[2 * exclude_sweep_size + 1, 1],
+            stride=1,
+            padding=[exclude_sweep_size, 0],
+        )
+        max_amps = max_amps[0, 0]
+        inds = inds[0, 0]
+        # torch `inds` gives loc of argmax at each position
+        # find those which actually *were* the max
+        unique_inds = inds.unique()
+        window_max_inds = unique_inds[inds.view(-1)[unique_inds] == unique_inds]
+
+        # voltage threshold (@Charlie, added by channel)
+        thresholds_by_channel = torch.as_tensor(
+                thresholds_torch[window_max_inds % num_channels], device=device, dtype=torch.float
+            )
+        max_amps_at_inds = max_amps.view(-1)[window_max_inds]
+        crossings = torch.nonzero(max_amps_at_inds > thresholds_by_channel).squeeze()
+        if not crossings.numel():
+            return np.array([]), np.array([]), np.array([])
+
+        # -- unravel the spike index
+        # (right now the indices are into flattened recording)
+        peak_inds = window_max_inds[crossings]
+        sample_inds = torch.div(peak_inds, num_channels, rounding_mode="floor")
+        chan_inds = peak_inds % num_channels
+        amplitudes = max_amps_at_inds[crossings]
+
+        # we need this due to the padding in convolution
+        valid_inds = torch.nonzero(
+            (0 < sample_inds) & (sample_inds < traces.shape[0] - 1)
+        ).squeeze()
+        if not sample_inds.numel():
+            return np.array([]), np.array([]), np.array([])
+        sample_inds = sample_inds[valid_inds]
+        chan_inds = chan_inds[valid_inds]
+        amplitudes = amplitudes[valid_inds]
+
+        # -- deduplication
+        # We deduplicate if the channel index is provided.
+        if neighbours_mask is not None:
+            neighbours_mask = torch.tensor(
+                neighbours_mask, device=device, dtype=torch.long
+            )
+
+            # -- temporal max pool
+            # still not sure why we can't just use `max_amps` instead of making
+            # this sparsely populated array, but it leads to a different result.
+            max_amps[:] = 0
+            max_amps[sample_inds, chan_inds] = amplitudes
+
+            # TODO: @Charlie
+            # 1. is this to deduplicate over neighbours using a wider window? Assuming propagation
+            # 2. can we derive max_window from exclude_sweep_ms (like max_window = 2*exclude_window_ms)?
```

Comment:
Yes, definitely. The units here are in samples, and it's the radius of the window rather than the diameter (i.e., no peaks will be as close as this). Is `exclude_window_ms` a radius or diameter?

Reply:
I think it's one sided, so maybe using `2*exclude_sweep_size` in the previous max pooling (instead of `order`) and skip this will do the job :) in the default case, this will be 6 (so something in between default `order` and `max_window`)

Reply:
Yup, I think that will work. I'll run some tests on my end to make sure the result is the same when we combine the steps.

Reply:
Thanks!

Reply:
OK, so, it turns out that it's important to have both of the temporal max poolings. However, we can combine the parameters `exclude_sweep_size` and `max_window` -- these can be set to the same number.

Reply:
(Sorry for deleting a comment, if you noticed it.)

It turns out that both max poolings are necessary, and the meaning of `max_window` and `exclude_sweep_size` are different.

The first max pooling, with radius `exclude_sweep_size` here, determines what can count as a peak. A voltage needs to be larger than `exclude_sweep_size` neighbors on both sides.

The second max pooling deduplicates these peaks: a peak must be the largest one within `max_window` timesteps.

`exclude_sweep_size` should be set smaller than `max_window`, since the first max pooling is basically finding spikes, and the second max pooling is picking out the biggest one in a larger neighborhood.

Reply:
I see, but then shouldn't the second pass be done also in case `neighbours_mask` is None? (i.e. by_channel detection)?

Reply:
Ok! Maybe let's discuss details over a call ;)